logger/src/dal.rs

```diff
@@ -0,0 +1,156 @@
+use std::time::SystemTime;
+
+use async_broadcast::{broadcast, Sender};
+use async_trait::async_trait;
+use chrono::NaiveDateTime;
+use prost_types::Timestamp;
+use shuttle_proto::logger::{LogItem, LogLine};
+use sqlx::{
+    migrate::Migrator,
+    postgres::PgConnectOptions,
+    types::chrono::{DateTime, Utc},
+    FromRow, PgPool, QueryBuilder,
+};
+use thiserror::Error;
+use tracing::error;
+
+use tonic::transport::Uri;
+
+pub static MIGRATIONS: Migrator = sqlx::migrate!("./migrations");
+
+#[derive(Error, Debug)]
+pub enum DalError {
+    #[error("database request failed: {0}")]
+    Sqlx(#[from] sqlx::Error),
+}
+
+#[async_trait]
+pub trait Dal {
+    /// Get logs for a deployment
+    async fn get_logs(&self, deployment_id: String) -> Result<Vec<Log>, DalError>;
+}
+
+#[derive(Clone)]
+pub struct Postgres {
+    pool: PgPool,
+    tx: Sender<Vec<Log>>,
+}
+
+impl Postgres {
+    /// This function creates all necessary tables and sets up a database connection pool.
+    pub async fn new(connection_uri: &Uri) -> Self {
+        let pool = PgPool::connect(connection_uri.to_string().as_str())
+            .await
+            .expect("to be able to connect to the postgres db using the connection url");
+        Self::from_pool(pool).await
+    }
+
+    pub async fn with_options(options: PgConnectOptions) -> Self {
+        let pool = PgPool::connect_with(options)
+            .await
+            .expect("to be able to connect to the postgres db using the pg connect options");
+        Self::from_pool(pool).await
+    }
+
+    async fn from_pool(pool: PgPool) -> Self {
+        MIGRATIONS
+            .run(&pool)
+            .await
+            .expect("to run migrations successfully");
+
+        // TODO: we switched to async_broadcast to resolve the infinite loop bug, but it wasn't related.
+        // Should we switch back to tokio::broadcast?
```

Comment:
We should address this comment in any way we want :smile:

Reply:
Reverted to tokio broadcast.